src/smr.rs

```diff
@@ -0,0 +1,1360 @@
+use std::collections::{HashMap, HashSet};
+use std::marker::PhantomData;
+use std::sync::Arc;
+use std::time::{Duration, Instant};
+
+use creep::Context;
+use futures::channel::mpsc::{unbounded, UnboundedReceiver, UnboundedSender};
+use futures::{select, StreamExt, TryFutureExt};
+use log::{debug, error, info, warn};
+
+use crate::auth::{AuthCell, AuthFixedConfig, AuthManage};
+use crate::cabinet::{Cabinet, Capsule};
+use crate::error::{ErrorInfo, ErrorKind};
+use crate::exec::ExecRequest;
+use crate::state::{ProposePrepare, Stage, StateInfo, Step};
+use crate::sync::{
+    Sync, SyncStat, BLOCK_BATCH, CLEAR_TIMEOUT_RATIO, HEIGHT_RATIO, SYNC_TIMEOUT_RATIO,
+};
+use crate::timeout::{TimeoutEvent, TimeoutInfo};
+use crate::types::{
+    Choke, ChokeQC, FetchedFullBlock, PreCommitQC, PreVoteQC, Proposal, SignedChoke, SignedHeight,
+    SignedPreCommit, SignedPreVote, SignedProposal, SyncRequest, SyncResponse, UpdateFrom, Vote,
+};
+use crate::{
+    Adapter, Address, Blk, ExecResult, Hash, Height, HeightRange, OverlordError, OverlordMsg,
+    OverlordResult, Proof, Round, St, TimeConfig, TinyHex, Wal, INIT_ROUND,
+};
+
+const POWER_CAP: u32 = 5;
+const TIME_DIVISOR: u64 = 10;
+
+const HEIGHT_WINDOW: Height = 5;
+const ROUND_WINDOW: Round = 5;
+
+pub type WrappedOverlordMsg<B> = (Context, OverlordMsg<B>);
+
+/// State Machine Replica
+pub struct SMR<A: Adapter<B, S>, B: Blk, S: St> {
+    address: Address,
+    state:   StateInfo<B>,
+    prepare: ProposePrepare<S>,
+    sync:    Sync<B>,
+
+    adapter: Arc<A>,
+    wal:     Wal,
+    cabinet: Cabinet<B>,
+    auth:    AuthManage<A, B, S>,
+    agent:   EventAgent<A, B, S>,
+
+    phantom_s: PhantomData<S>,
+}
+
+impl<A, B, S> SMR<A, B, S>
+where
+    A: Adapter<B, S>,
+    B: Blk,
+    S: St,
+{
+    pub async fn new(
+        auth_fixed_config: AuthFixedConfig,
+        adapter: &Arc<A>,
+        from_net: UnboundedReceiver<(Context, OverlordMsg<B>)>,
+        to_net: UnboundedSender<(Context, OverlordMsg<B>)>,
+        from_exec: UnboundedReceiver<ExecResult<S>>,
+        to_exec: UnboundedSender<ExecRequest<S>>,
+        wal_path: &str,
+    ) -> Self {
+        let address = &auth_fixed_config.address.clone();
+
+        let wal = Wal::new(wal_path);
+        let rst = wal.load_state();
+
+        let state;
+        let from;
+        if let Err(e) = rst {
+            warn!("Load state from wal failed: {}! Try to recover state by the adapter, which face security risk if majority auth nodes lost their wal file at the same time", e);
```

Comment:
I think this should be handled in overlord. Wal writing may be haltted in a sudden shutdown, which will make wal load failed. This happens not frequency but will happen. So, as a backup, load from storage is a basic demand for overlord.

Reply:
I mean `wal.load_state` should return `Result<Option<StateInfo>>`, `State Not Found` is normal behavior (such as a new chain), which should not be mixed with other errors.

Reply:
It's just designed for simple. Because in wal, I have a safe_load_file function, which will always make sure there's a wal file, so the only error is decode error.

Reply:
And, It's not that necessary to distinguish these two cases.